### PR TITLE
Add block write!/read! ability to allow for better coding style for bit manipulations.

### PR DESCRIPTION
**Pull request by chrisnappi**
This pull request intends to add a method to provide for better coding style when dealing with large numbers of bit manipulations within a register.  These manipulations can be hard to read, and error prone when '!' is in play.

Previous code style:
```
  $dut.mtr.mct.reg(:bstart).bits(:bstart).write(0x4) # Program BIST but don't start them
  $dut.mtr.mct.reg(:bstart).bits(:cens).write(1) 
  $dut.mtr.mct.reg(:bstart).bits(:cenac).write(1)
  $dut.mtr.mct.reg(:bstart).bits(:brst).write(1)
  $dut.mtr.mct.reg(:bstart).bits(:cof).write(1)
  $dut.mtr.mct.reg(:bstart).bits(:clken).write(1)
  $dut.mtr.mct.reg(:bstart).write!
```
New coding style:
```
  $dut.wreg($dut.mtr.mct.reg(:bstart)) do |r|
    r.bits(:cens).write(1) 
    r.bits(:cenac).write(1)
    r.bits(:brst).write(1)
    r.bits(:cof).write(1)
    r.bits(:clken).write(1)
  end
  $dut.mtr.mct.reg(:bstart).write!
```

### CODE REVIEW

**Comment by ginty** [comment]
Like the idea, would this not be even better?

~~~ruby
$dut.mtr.mct.reg(:bstart).write! do |r|
    r.bits(:cens).write(1) 
    r.bits(:cenac).write(1)
    r.bits(:brst).write(1)
    r.bits(:cof).write(1)
    r.bits(:clken).write(1)
end
~~~

**Comment by chrisnappi** [comment]
Yes, that would be better.  Not sure how to implement :)

**Comment by ginty** [comment]
Not much different, this is the write! method (I guess you might also do read! while you are at it) - https://github.com/Origen-SDK/origen/blob/master/lib/origen/registers/bit_collection.rb#L509

Something like this should do it:

~~~ruby
yield self if block_given?
# Original code here
~~~

File: lib/origen/registers/bit_collection.rb [hunk 1]
```diff
@@ -509,6 +509,7 @@ def overlay_str
       def write!(value = nil, options = {})
         value, options = nil, value if value.is_a?(Hash)
         write(value, options) if value
+        yield @reg if block_given?
```
Comment (ginty) [inline lib/origen/registers/bit_collection.rb]: Think these should be yield self and not @reg.

They are effectively the same when operating on a register, but this update does also support the same kind of operation on a subset of bits:

~~~ruby
some_reg.some_bits.write! do |b|
  b[1].write(1)
  b[4].write(1)
end
~~~

By yielding @reg the internal object would be the register and not the subset of bits and any relative position references like above would be off.

Comment (chrisnappi) [reply lib/origen/registers/bit_collection.rb]: Hmm - that does not work...

Comment (chrisnappi) [reply lib/origen/registers/bit_collection.rb]: TypeError: no implicit conversion of Symbol into Integer
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/registers/bit_collection.rb:48:in `fetch'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/registers/bit_collection.rb:48:in `block in []'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/registers/bit_collection.rb:47:in `each'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/registers/bit_collection.rb:47:in `[]'
	from (origen):5:in `block in irb_binding'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/registers/bit_collection.rb:512:in `write!'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/registers/reg.rb:1228:in `method_missing'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/registers.rb:249:in `method_missing'
	from (origen):4
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/commands/interactive.rb:79:in `<module:Origen>'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/commands/interactive.rb:11:in `<top (required)>'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/commands.rb:233:in `require'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/lib/origen/commands.rb:233:in `<top (required)>'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/bin/origen:172:in `require'
	from /proj/mcusoc_kodiak/users/chrisn/origen/origen/bin/origen:172:in `<top (required)>'
	from ./lbin/origen:17:in `load'

Comment (ginty) [reply lib/origen/registers/bit_collection.rb]: The example as I wrote it above you mean, or something else?

Comment (chrisnappi) [reply lib/origen/registers/bit_collection.rb]: Using self instead of @reg results in error above

Comment (ginty) [reply lib/origen/registers/bit_collection.rb]: Just pushed an update that fixes it.
I though `<bit collection>.bits` worked, but it has to be `<reg>.bits`

**Comment by chrisnappi** [comment]
Excellent - thanks!

**Comment by chrisnappi** [comment]
Yikes - this is generating incorrect output (the @reg version worked right).  Doing more testing to determine the issue.

**Pull request merged**